File: backend/src/application/service/validator/problem.ts
import * as Joi from 'joi'

const SOURCE_OPTIONS: string[] = [
  'gpt-4',
  'gpt-3',
  'model',
  'uue',
  'book',
  'extreme',
]
const DIFFICULTY_OPTIONS: string[] = ['normal', 'easy', 'medium', 'hard']
const TARGET_OPTIONS: string[] = [
  'elementary',
  'university',
  'highschool',
  'general',
]
const CONTENT_OPTIONS = ['text', 'image', 'equation', 'video']
const ANSWER_TYPE_OPTIONS: string[] = ['options', 'short', 'trueFalse']
const COURSES_OPTIONS: string[] = [
  'mathematics',
  'biology',
  'physics',
  'chemistry',
  'history',
  'geography',
  'civics',
  'economics',
  'amharic',
  'english',
  'civics',
  'computer',
  'general',
]

export function validateProblem(problem: any) {
  const schema = Joi.object({
    published: Joi.boolean().required(),
    source: Joi.object({
      name: Joi.string()
        .valid(...SOURCE_OPTIONS)
        .required(),
      value: Joi.string(),
      year: Joi.number(),
    }).required(),
    details: Joi.object({
      target: Joi.string()
        .valid(...TARGET_OPTIONS)
        .required(),
      grade: Joi.string(),
      unit: Joi.string(),
      courses: Joi.string()
        .valid(...COURSES_OPTIONS)
        .required(),
      topic: Joi.string(),
      difficulty: Joi.string().valid(...DIFFICULTY_OPTIONS),
    }).required(),
    question: Joi.array().items(
      Joi.object({
        type: Joi.string()
          .valid(...CONTENT_OPTIONS)
          .required(),
        value: Joi.string().required(),
      }),
    ),
    point: Joi.object({
      correct: Joi.number().required(),
      wrong: Joi.number().required(),
    }).required(),
    answer: Joi.object({
      type: Joi.string()
        .valid(...ANSWER_TYPE_OPTIONS)
        .required(),
      options: Joi.object({
        choice: Joi.array().items(
          Joi.object({
            correct: Joi.boolean().required(),
            data: Joi.array()
              .items(
                Joi.object({
                  type: Joi.string()
                    .valid(...CONTENT_OPTIONS)
                    .required(),
                  value: Joi.string().required(),
                }),
              )
              .required(),
          }),
        ),
      }),
      short: Joi.object({
        correct: Joi.string(),
      }),
      trueFalse: Joi.object({
        correct: Joi.string(),
      }),
      explanation: Joi.array().items(
        Joi.object({
          type: Joi.string().valid(...CONTENT_OPTIONS),
          value: Joi.string(),
        }),
      ),
    }).required(),
    essay: Joi.string().regex(/^[0-9a-fA-F]{24}$/),
    createdAt: Joi.date(),
    updatedAt: Joi.date(),
  })

  return schema.validate(problem)
}

export default validateProblem

File: mobile/lib/features/ed_ai/data/models/problem.dart
import 'dart:convert';

import 'package:mobile/features/ed_ai/data/models/answer.dart';
import 'package:mobile/features/ed_ai/data/models/content.dart';
import 'package:mobile/features/ed_ai/domains/entities/answer.dart';
import 'package:mobile/features/ed_ai/domains/entities/content.dart';
import 'package:mobile/features/ed_ai/domains/entities/problem.dart';

class ProblemModel extends Problem {
  const ProblemModel({
    required String id,
    required String source,
    required String target,
    required String courses,
    required List<Content> question,
    required Answer answer,
    required int correctPoint,
    required int wrongPoint,
    String? essayId,
    String? difficulty,
    String? topic,
    String? unit,
    String? grade,
    String? value,
    int? year,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) : super(
          id: id,
          source: source,
          target: target,
          courses: courses,
          question: question,
          answer: answer,
          essayId: essayId,
          difficulty: difficulty,
          topic: topic,
          unit: unit,
          grade: grade,
          correctPoint: correctPoint,
          wrongPoint: wrongPoint,
          value: value,
          year: year,
          createdAt: createdAt,
          updatedAt: updatedAt,
        );

  factory ProblemModel.fromJson(Map<String, dynamic> json) {
    return ProblemModel(
      id: json['_id'] ?? '',
      source: json['source'] != null ? json['source']['name'] ?? '' : '',
      target: json['details'] != null ? json['details']['target'] ?? '' : '',
      courses: json['details'] != null ? json['details']['courses'] ?? '' : '',
      question: (json['question'] as List?)
              ?.map((e) => ContentModel.fromJson(e ?? {}))
              .toList() ??
          [],
      answer: AnswerModel.fromJson(json['answer'] ?? {}),
      essayId: json['essayId'] ?? '',
      difficulty:
          json['details'] != null ? json['details']['difficulty'] ?? '' : '',
      topic: json['details'] != null ? json['details']['topic'] ?? '' : '',
      unit: json['details'] != null ? json['details']['unit'] ?? '' : '',
      grade: json['details'] != null ? json['details']['grade'] ?? '' : '',
      correctPoint: json['point'] != null ? json['point']['correct'] ?? 0 : 0,
      wrongPoint: json['point'] != null ? json['point']['wrong'] ?? 0 : 0,
      value: json['source'] != null ? json['source']['value'] ?? '' : '',
      year: json['source'] != null ? json['source']['year'] ?? 0 : 0,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toDbJson() {
    return {
      'id': id,
      'source': source,
      'target': target,
      'courses': courses,
      'question': jsonEncode(question),
      'answer': jsonEncode(answer),
      'essayId': essayId,
      'difficulty': difficulty,
      'topic': topic,
      'unit': unit,
      'grade': grade,
      'correctPoint': correctPoint,
      'wrongPoint': wrongPoint,
      'value': value,
      'year': year,
      'createdAt': createdAt.toString(),
      'updatedAt': updatedAt.toString(),
    };
  }

  ProblemModel.fromDbJson(Map<String, dynamic> json)
      : this(
          id: json['id'],
          source: json['source'],
          target: json['target'],
          courses: json['courses'],
          question: (jsonDecode(json['question']) as List)
              .map((e) => ContentModel.fromJson(e))
              .toList(),
          answer: AnswerModel.fromJson(jsonDecode(json['answer'])),
          essayId: json['essayId'],
          difficulty: json['difficulty'],
          topic: json['topic'],
          unit: json['unit'],
          grade: json['grade'],
          correctPoint: json['correctPoint'],
          wrongPoint: json['wrongPoint'],
          value: json['value'],
          year: json['year'],
          createdAt: DateTime.parse(json['createdAt']),
          updatedAt: DateTime.parse(json['updatedAt']),
        );
}
